File: domain/src/main/java/SagaBase.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public abstract class SagaBase implements Saga {
    private final IRouteEvents _eventRouter;
    private final List<BaseEvent> _uncommitted = new ArrayList<>();
    private final List<DomainCommand> _undispatched = new ArrayList<>();
    private final String _id;
    private Integer _version;

    protected SagaBase(String id, IRouteEvents eventRouter) {
        _eventRouter = eventRouter;
        _id = id;
    }

    public void Transition(BaseEvent message) {

        _eventRouter.dispatch(message);

        _uncommitted.add(message);
        ++_version;
    }

    @Override
    public Iterable<Object> getUncommittedEvents() {
        return Arrays.asList(_uncommitted.toArray());
    }

    @Override
    public void clearUncommittedEvents() {
        _uncommitted.clear();
    }

    @Override
    public Iterable<Object> getUndispatchedMessages() {
        return Arrays.asList(_undispatched.toArray());
    }

    @Override
    public void clearUndispatchedMessages() {
        _undispatched.clear();
    }

    protected void dispatch(DomainCommand message){
        _undispatched.add(message);
    }

    protected void register(Consumer<BaseEvent> handler){
        _eventRouter.register(handler);
    }
}

File: modules/domain/src/main/kotlin/org/openmedstack/domain/SagaBase.kt
package org.openmedstack.domain

import org.openmedstack.commands.DomainCommand
import org.openmedstack.events.BaseEvent
import java.util.*
import java.util.function.Consumer

abstract class SagaBase protected constructor(id: String, eventRouter: IRouteEvents?) : Saga {
    private val _eventRouter: IRouteEvents
    private val _uncommitted: ArrayList<BaseEvent> = ArrayList()
    private val _undispatched: ArrayList<DomainCommand> = ArrayList()
    private val _id: String
    private var _version = 0

    override val id: String
        get() = _id

    override val version: Int
        get() = _version

    @Throws(HandlerForDomainEventNotFoundException::class)
    override fun transition(message: BaseEvent) {
        _eventRouter.dispatch(message, message.javaClass)
        _uncommitted.add(message)
        _version += 1
    }

    override val uncommittedEvents: Iterable<BaseEvent>
        get() = _uncommitted

    override fun clearUncommittedEvents() {
        _uncommitted.clear()
    }

    override val undispatchedMessages: Iterable<Any>
        get() = listOf<Any>(_undispatched)

    override fun clearUndispatchedMessages() {
        _undispatched.clear()
    }

    protected fun dispatch(message: DomainCommand) {
        _undispatched.add(message)
    }

    protected fun register(handler: Consumer<BaseEvent>) {
        _eventRouter.register(handler)
    }

    init {
        _eventRouter = eventRouter ?: ConventionEventRouter(true, this)
        _id = id
    }
}

